{"ast":null,"code":"/*!\r\n * @overview es6-promise - a tiny implementation of Promises/A+.\r\n * @copyright Copyright (c) 2014 Yehuda Katz, Tom Dale, Stefan Penner and contributors (Conversion to ES6 API by Jake Archibald)\r\n * @license   Licensed under MIT license\r\n *            See https://raw.githubusercontent.com/stefanpenner/es6-promise/master/LICENSE\r\n * @version   v4.2.8+1e68dce6\r\n */\n(function (global, factory) {\n  typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() : typeof define === 'function' && define.amd ? define(factory) : global.ES6Promise = factory();\n})(this, function () {\n  'use strict';\n\n  function objectOrFunction(x) {\n    var type = typeof x;\n    return x !== null && (type === 'object' || type === 'function');\n  }\n\n  function isFunction(x) {\n    return typeof x === 'function';\n  }\n\n  var _isArray = void 0;\n\n  if (Array.isArray) {\n    _isArray = Array.isArray;\n  } else {\n    _isArray = function (x) {\n      return Object.prototype.toString.call(x) === '[object Array]';\n    };\n  }\n\n  var isArray = _isArray;\n  var len = 0;\n  var vertxNext = void 0;\n  var customSchedulerFn = void 0;\n\n  var asap = function asap(callback, arg) {\n    queue[len] = callback;\n    queue[len + 1] = arg;\n    len += 2;\n\n    if (len === 2) {\n      // If len is 2, that means that we need to schedule an async flush.\n      // If additional callbacks are queued before the queue is flushed, they\n      // will be processed by this flush that we are scheduling.\n      if (customSchedulerFn) {\n        customSchedulerFn(flush);\n      } else {\n        scheduleFlush();\n      }\n    }\n  };\n\n  function setScheduler(scheduleFn) {\n    customSchedulerFn = scheduleFn;\n  }\n\n  function setAsap(asapFn) {\n    asap = asapFn;\n  }\n\n  var browserWindow = typeof window !== 'undefined' ? window : undefined;\n  var browserGlobal = browserWindow || {};\n  var BrowserMutationObserver = browserGlobal.MutationObserver || browserGlobal.WebKitMutationObserver;\n  var isNode = typeof self === 'undefined' && typeof process !== 'undefined' && {}.toString.call(process) === '[object process]'; // test for web worker but not in IE10\n\n  var isWorker = typeof Uint8ClampedArray !== 'undefined' && typeof importScripts !== 'undefined' && typeof MessageChannel !== 'undefined'; // node\n\n  function useNextTick() {\n    // node version 0.10.x displays a deprecation warning when nextTick is used recursively\n    // see https://github.com/cujojs/when/issues/410 for details\n    return function () {\n      return process.nextTick(flush);\n    };\n  } // vertx\n\n\n  function useVertxTimer() {\n    if (typeof vertxNext !== 'undefined') {\n      return function () {\n        vertxNext(flush);\n      };\n    }\n\n    return useSetTimeout();\n  }\n\n  function useMutationObserver() {\n    var iterations = 0;\n    var observer = new BrowserMutationObserver(flush);\n    var node = document.createTextNode('');\n    observer.observe(node, {\n      characterData: true\n    });\n    return function () {\n      node.data = iterations = ++iterations % 2;\n    };\n  } // web worker\n\n\n  function useMessageChannel() {\n    var channel = new MessageChannel();\n    channel.port1.onmessage = flush;\n    return function () {\n      return channel.port2.postMessage(0);\n    };\n  }\n\n  function useSetTimeout() {\n    // Store setTimeout reference so es6-promise will be unaffected by\n    // other code modifying setTimeout (like sinon.useFakeTimers())\n    var globalSetTimeout = setTimeout;\n    return function () {\n      return globalSetTimeout(flush, 1);\n    };\n  }\n\n  var queue = new Array(1000);\n\n  function flush() {\n    for (var i = 0; i < len; i += 2) {\n      var callback = queue[i];\n      var arg = queue[i + 1];\n      callback(arg);\n      queue[i] = undefined;\n      queue[i + 1] = undefined;\n    }\n\n    len = 0;\n  }\n\n  function attemptVertx() {\n    try {\n      var vertx = Function('return this')().require('vertx');\n\n      vertxNext = vertx.runOnLoop || vertx.runOnContext;\n      return useVertxTimer();\n    } catch (e) {\n      return useSetTimeout();\n    }\n  }\n\n  var scheduleFlush = void 0; // Decide what async method to use to triggering processing of queued callbacks:\n\n  if (isNode) {\n    scheduleFlush = useNextTick();\n  } else if (BrowserMutationObserver) {\n    scheduleFlush = useMutationObserver();\n  } else if (isWorker) {\n    scheduleFlush = useMessageChannel();\n  } else if (browserWindow === undefined && typeof require === 'function') {\n    scheduleFlush = attemptVertx();\n  } else {\n    scheduleFlush = useSetTimeout();\n  }\n\n  function then(onFulfillment, onRejection) {\n    var parent = this;\n    var child = new this.constructor(noop);\n\n    if (child[PROMISE_ID] === undefined) {\n      makePromise(child);\n    }\n\n    var _state = parent._state;\n\n    if (_state) {\n      var callback = arguments[_state - 1];\n      asap(function () {\n        return invokeCallback(_state, child, callback, parent._result);\n      });\n    } else {\n      subscribe(parent, child, onFulfillment, onRejection);\n    }\n\n    return child;\n  }\n  /**\r\n    `Promise.resolve` returns a promise that will become resolved with the\r\n    passed `value`. It is shorthand for the following:\r\n  \r\n    ```javascript\r\n    let promise = new Promise(function(resolve, reject){\r\n      resolve(1);\r\n    });\r\n  \r\n    promise.then(function(value){\r\n      // value === 1\r\n    });\r\n    ```\r\n  \r\n    Instead of writing the above, your code now simply becomes the following:\r\n  \r\n    ```javascript\r\n    let promise = Promise.resolve(1);\r\n  \r\n    promise.then(function(value){\r\n      // value === 1\r\n    });\r\n    ```\r\n  \r\n    @method resolve\r\n    @static\r\n    @param {Any} value value that the returned promise will be resolved with\r\n    Useful for tooling.\r\n    @return {Promise} a promise that will become fulfilled with the given\r\n    `value`\r\n  */\n\n\n  function resolve$1(object) {\n    /*jshint validthis:true */\n    var Constructor = this;\n\n    if (object && typeof object === 'object' && object.constructor === Constructor) {\n      return object;\n    }\n\n    var promise = new Constructor(noop);\n    resolve(promise, object);\n    return promise;\n  }\n\n  var PROMISE_ID = Math.random().toString(36).substring(2);\n\n  function noop() {}\n\n  var PENDING = void 0;\n  var FULFILLED = 1;\n  var REJECTED = 2;\n\n  function selfFulfillment() {\n    return new TypeError(\"You cannot resolve a promise with itself\");\n  }\n\n  function cannotReturnOwn() {\n    return new TypeError('A promises callback cannot return that same promise.');\n  }\n\n  function tryThen(then$$1, value, fulfillmentHandler, rejectionHandler) {\n    try {\n      then$$1.call(value, fulfillmentHandler, rejectionHandler);\n    } catch (e) {\n      return e;\n    }\n  }\n\n  function handleForeignThenable(promise, thenable, then$$1) {\n    asap(function (promise) {\n      var sealed = false;\n      var error = tryThen(then$$1, thenable, function (value) {\n        if (sealed) {\n          return;\n        }\n\n        sealed = true;\n\n        if (thenable !== value) {\n          resolve(promise, value);\n        } else {\n          fulfill(promise, value);\n        }\n      }, function (reason) {\n        if (sealed) {\n          return;\n        }\n\n        sealed = true;\n        reject(promise, reason);\n      }, 'Settle: ' + (promise._label || ' unknown promise'));\n\n      if (!sealed && error) {\n        sealed = true;\n        reject(promise, error);\n      }\n    }, promise);\n  }\n\n  function handleOwnThenable(promise, thenable) {\n    if (thenable._state === FULFILLED) {\n      fulfill(promise, thenable._result);\n    } else if (thenable._state === REJECTED) {\n      reject(promise, thenable._result);\n    } else {\n      subscribe(thenable, undefined, function (value) {\n        return resolve(promise, value);\n      }, function (reason) {\n        return reject(promise, reason);\n      });\n    }\n  }\n\n  function handleMaybeThenable(promise, maybeThenable, then$$1) {\n    if (maybeThenable.constructor === promise.constructor && then$$1 === then && maybeThenable.constructor.resolve === resolve$1) {\n      handleOwnThenable(promise, maybeThenable);\n    } else {\n      if (then$$1 === undefined) {\n        fulfill(promise, maybeThenable);\n      } else if (isFunction(then$$1)) {\n        handleForeignThenable(promise, maybeThenable, then$$1);\n      } else {\n        fulfill(promise, maybeThenable);\n      }\n    }\n  }\n\n  function resolve(promise, value) {\n    if (promise === value) {\n      reject(promise, selfFulfillment());\n    } else if (objectOrFunction(value)) {\n      var then$$1 = void 0;\n\n      try {\n        then$$1 = value.then;\n      } catch (error) {\n        reject(promise, error);\n        return;\n      }\n\n      handleMaybeThenable(promise, value, then$$1);\n    } else {\n      fulfill(promise, value);\n    }\n  }\n\n  function publishRejection(promise) {\n    if (promise._onerror) {\n      promise._onerror(promise._result);\n    }\n\n    publish(promise);\n  }\n\n  function fulfill(promise, value) {\n    if (promise._state !== PENDING) {\n      return;\n    }\n\n    promise._result = value;\n    promise._state = FULFILLED;\n\n    if (promise._subscribers.length !== 0) {\n      asap(publish, promise);\n    }\n  }\n\n  function reject(promise, reason) {\n    if (promise._state !== PENDING) {\n      return;\n    }\n\n    promise._state = REJECTED;\n    promise._result = reason;\n    asap(publishRejection, promise);\n  }\n\n  function subscribe(parent, child, onFulfillment, onRejection) {\n    var _subscribers = parent._subscribers;\n    var length = _subscribers.length;\n    parent._onerror = null;\n    _subscribers[length] = child;\n    _subscribers[length + FULFILLED] = onFulfillment;\n    _subscribers[length + REJECTED] = onRejection;\n\n    if (length === 0 && parent._state) {\n      asap(publish, parent);\n    }\n  }\n\n  function publish(promise) {\n    var subscribers = promise._subscribers;\n    var settled = promise._state;\n\n    if (subscribers.length === 0) {\n      return;\n    }\n\n    var child = void 0,\n        callback = void 0,\n        detail = promise._result;\n\n    for (var i = 0; i < subscribers.length; i += 3) {\n      child = subscribers[i];\n      callback = subscribers[i + settled];\n\n      if (child) {\n        invokeCallback(settled, child, callback, detail);\n      } else {\n        callback(detail);\n      }\n    }\n\n    promise._subscribers.length = 0;\n  }\n\n  function invokeCallback(settled, promise, callback, detail) {\n    var hasCallback = isFunction(callback),\n        value = void 0,\n        error = void 0,\n        succeeded = true;\n\n    if (hasCallback) {\n      try {\n        value = callback(detail);\n      } catch (e) {\n        succeeded = false;\n        error = e;\n      }\n\n      if (promise === value) {\n        reject(promise, cannotReturnOwn());\n        return;\n      }\n    } else {\n      value = detail;\n    }\n\n    if (promise._state !== PENDING) {// noop\n    } else if (hasCallback && succeeded) {\n      resolve(promise, value);\n    } else if (succeeded === false) {\n      reject(promise, error);\n    } else if (settled === FULFILLED) {\n      fulfill(promise, value);\n    } else if (settled === REJECTED) {\n      reject(promise, value);\n    }\n  }\n\n  function initializePromise(promise, resolver) {\n    try {\n      resolver(function resolvePromise(value) {\n        resolve(promise, value);\n      }, function rejectPromise(reason) {\n        reject(promise, reason);\n      });\n    } catch (e) {\n      reject(promise, e);\n    }\n  }\n\n  var id = 0;\n\n  function nextId() {\n    return id++;\n  }\n\n  function makePromise(promise) {\n    promise[PROMISE_ID] = id++;\n    promise._state = undefined;\n    promise._result = undefined;\n    promise._subscribers = [];\n  }\n\n  function validationError() {\n    return new Error('Array Methods must be provided an Array');\n  }\n\n  var Enumerator = function () {\n    function Enumerator(Constructor, input) {\n      this._instanceConstructor = Constructor;\n      this.promise = new Constructor(noop);\n\n      if (!this.promise[PROMISE_ID]) {\n        makePromise(this.promise);\n      }\n\n      if (isArray(input)) {\n        this.length = input.length;\n        this._remaining = input.length;\n        this._result = new Array(this.length);\n\n        if (this.length === 0) {\n          fulfill(this.promise, this._result);\n        } else {\n          this.length = this.length || 0;\n\n          this._enumerate(input);\n\n          if (this._remaining === 0) {\n            fulfill(this.promise, this._result);\n          }\n        }\n      } else {\n        reject(this.promise, validationError());\n      }\n    }\n\n    Enumerator.prototype._enumerate = function _enumerate(input) {\n      for (var i = 0; this._state === PENDING && i < input.length; i++) {\n        this._eachEntry(input[i], i);\n      }\n    };\n\n    Enumerator.prototype._eachEntry = function _eachEntry(entry, i) {\n      var c = this._instanceConstructor;\n      var resolve$$1 = c.resolve;\n\n      if (resolve$$1 === resolve$1) {\n        var _then = void 0;\n\n        var error = void 0;\n        var didError = false;\n\n        try {\n          _then = entry.then;\n        } catch (e) {\n          didError = true;\n          error = e;\n        }\n\n        if (_then === then && entry._state !== PENDING) {\n          this._settledAt(entry._state, i, entry._result);\n        } else if (typeof _then !== 'function') {\n          this._remaining--;\n          this._result[i] = entry;\n        } else if (c === Promise$1) {\n          var promise = new c(noop);\n\n          if (didError) {\n            reject(promise, error);\n          } else {\n            handleMaybeThenable(promise, entry, _then);\n          }\n\n          this._willSettleAt(promise, i);\n        } else {\n          this._willSettleAt(new c(function (resolve$$1) {\n            return resolve$$1(entry);\n          }), i);\n        }\n      } else {\n        this._willSettleAt(resolve$$1(entry), i);\n      }\n    };\n\n    Enumerator.prototype._settledAt = function _settledAt(state, i, value) {\n      var promise = this.promise;\n\n      if (promise._state === PENDING) {\n        this._remaining--;\n\n        if (state === REJECTED) {\n          reject(promise, value);\n        } else {\n          this._result[i] = value;\n        }\n      }\n\n      if (this._remaining === 0) {\n        fulfill(promise, this._result);\n      }\n    };\n\n    Enumerator.prototype._willSettleAt = function _willSettleAt(promise, i) {\n      var enumerator = this;\n      subscribe(promise, undefined, function (value) {\n        return enumerator._settledAt(FULFILLED, i, value);\n      }, function (reason) {\n        return enumerator._settledAt(REJECTED, i, reason);\n      });\n    };\n\n    return Enumerator;\n  }();\n  /**\r\n    `Promise.all` accepts an array of promises, and returns a new promise which\r\n    is fulfilled with an array of fulfillment values for the passed promises, or\r\n    rejected with the reason of the first passed promise to be rejected. It casts all\r\n    elements of the passed iterable to promises as it runs this algorithm.\r\n  \r\n    Example:\r\n  \r\n    ```javascript\r\n    let promise1 = resolve(1);\r\n    let promise2 = resolve(2);\r\n    let promise3 = resolve(3);\r\n    let promises = [ promise1, promise2, promise3 ];\r\n  \r\n    Promise.all(promises).then(function(array){\r\n      // The array here would be [ 1, 2, 3 ];\r\n    });\r\n    ```\r\n  \r\n    If any of the `promises` given to `all` are rejected, the first promise\r\n    that is rejected will be given as an argument to the returned promises's\r\n    rejection handler. For example:\r\n  \r\n    Example:\r\n  \r\n    ```javascript\r\n    let promise1 = resolve(1);\r\n    let promise2 = reject(new Error(\"2\"));\r\n    let promise3 = reject(new Error(\"3\"));\r\n    let promises = [ promise1, promise2, promise3 ];\r\n  \r\n    Promise.all(promises).then(function(array){\r\n      // Code here never runs because there are rejected promises!\r\n    }, function(error) {\r\n      // error.message === \"2\"\r\n    });\r\n    ```\r\n  \r\n    @method all\r\n    @static\r\n    @param {Array} entries array of promises\r\n    @param {String} label optional string for labeling the promise.\r\n    Useful for tooling.\r\n    @return {Promise} promise that is fulfilled when all `promises` have been\r\n    fulfilled, or rejected if any of them become rejected.\r\n    @static\r\n  */\n\n\n  function all(entries) {\n    return new Enumerator(this, entries).promise;\n  }\n  /**\r\n    `Promise.race` returns a new promise which is settled in the same way as the\r\n    first passed promise to settle.\r\n  \r\n    Example:\r\n  \r\n    ```javascript\r\n    let promise1 = new Promise(function(resolve, reject){\r\n      setTimeout(function(){\r\n        resolve('promise 1');\r\n      }, 200);\r\n    });\r\n  \r\n    let promise2 = new Promise(function(resolve, reject){\r\n      setTimeout(function(){\r\n        resolve('promise 2');\r\n      }, 100);\r\n    });\r\n  \r\n    Promise.race([promise1, promise2]).then(function(result){\r\n      // result === 'promise 2' because it was resolved before promise1\r\n      // was resolved.\r\n    });\r\n    ```\r\n  \r\n    `Promise.race` is deterministic in that only the state of the first\r\n    settled promise matters. For example, even if other promises given to the\r\n    `promises` array argument are resolved, but the first settled promise has\r\n    become rejected before the other promises became fulfilled, the returned\r\n    promise will become rejected:\r\n  \r\n    ```javascript\r\n    let promise1 = new Promise(function(resolve, reject){\r\n      setTimeout(function(){\r\n        resolve('promise 1');\r\n      }, 200);\r\n    });\r\n  \r\n    let promise2 = new Promise(function(resolve, reject){\r\n      setTimeout(function(){\r\n        reject(new Error('promise 2'));\r\n      }, 100);\r\n    });\r\n  \r\n    Promise.race([promise1, promise2]).then(function(result){\r\n      // Code here never runs\r\n    }, function(reason){\r\n      // reason.message === 'promise 2' because promise 2 became rejected before\r\n      // promise 1 became fulfilled\r\n    });\r\n    ```\r\n  \r\n    An example real-world use case is implementing timeouts:\r\n  \r\n    ```javascript\r\n    Promise.race([ajax('foo.json'), timeout(5000)])\r\n    ```\r\n  \r\n    @method race\r\n    @static\r\n    @param {Array} promises array of promises to observe\r\n    Useful for tooling.\r\n    @return {Promise} a promise which settles in the same way as the first passed\r\n    promise to settle.\r\n  */\n\n\n  function race(entries) {\n    /*jshint validthis:true */\n    var Constructor = this;\n\n    if (!isArray(entries)) {\n      return new Constructor(function (_, reject) {\n        return reject(new TypeError('You must pass an array to race.'));\n      });\n    } else {\n      return new Constructor(function (resolve, reject) {\n        var length = entries.length;\n\n        for (var i = 0; i < length; i++) {\n          Constructor.resolve(entries[i]).then(resolve, reject);\n        }\n      });\n    }\n  }\n  /**\r\n    `Promise.reject` returns a promise rejected with the passed `reason`.\r\n    It is shorthand for the following:\r\n  \r\n    ```javascript\r\n    let promise = new Promise(function(resolve, reject){\r\n      reject(new Error('WHOOPS'));\r\n    });\r\n  \r\n    promise.then(function(value){\r\n      // Code here doesn't run because the promise is rejected!\r\n    }, function(reason){\r\n      // reason.message === 'WHOOPS'\r\n    });\r\n    ```\r\n  \r\n    Instead of writing the above, your code now simply becomes the following:\r\n  \r\n    ```javascript\r\n    let promise = Promise.reject(new Error('WHOOPS'));\r\n  \r\n    promise.then(function(value){\r\n      // Code here doesn't run because the promise is rejected!\r\n    }, function(reason){\r\n      // reason.message === 'WHOOPS'\r\n    });\r\n    ```\r\n  \r\n    @method reject\r\n    @static\r\n    @param {Any} reason value that the returned promise will be rejected with.\r\n    Useful for tooling.\r\n    @return {Promise} a promise rejected with the given `reason`.\r\n  */\n\n\n  function reject$1(reason) {\n    /*jshint validthis:true */\n    var Constructor = this;\n    var promise = new Constructor(noop);\n    reject(promise, reason);\n    return promise;\n  }\n\n  function needsResolver() {\n    throw new TypeError('You must pass a resolver function as the first argument to the promise constructor');\n  }\n\n  function needsNew() {\n    throw new TypeError(\"Failed to construct 'Promise': Please use the 'new' operator, this object constructor cannot be called as a function.\");\n  }\n  /**\r\n    Promise objects represent the eventual result of an asynchronous operation. The\r\n    primary way of interacting with a promise is through its `then` method, which\r\n    registers callbacks to receive either a promise's eventual value or the reason\r\n    why the promise cannot be fulfilled.\r\n  \r\n    Terminology\r\n    -----------\r\n  \r\n    - `promise` is an object or function with a `then` method whose behavior conforms to this specification.\r\n    - `thenable` is an object or function that defines a `then` method.\r\n    - `value` is any legal JavaScript value (including undefined, a thenable, or a promise).\r\n    - `exception` is a value that is thrown using the throw statement.\r\n    - `reason` is a value that indicates why a promise was rejected.\r\n    - `settled` the final resting state of a promise, fulfilled or rejected.\r\n  \r\n    A promise can be in one of three states: pending, fulfilled, or rejected.\r\n  \r\n    Promises that are fulfilled have a fulfillment value and are in the fulfilled\r\n    state.  Promises that are rejected have a rejection reason and are in the\r\n    rejected state.  A fulfillment value is never a thenable.\r\n  \r\n    Promises can also be said to *resolve* a value.  If this value is also a\r\n    promise, then the original promise's settled state will match the value's\r\n    settled state.  So a promise that *resolves* a promise that rejects will\r\n    itself reject, and a promise that *resolves* a promise that fulfills will\r\n    itself fulfill.\r\n  \r\n  \r\n    Basic Usage:\r\n    ------------\r\n  \r\n    ```js\r\n    let promise = new Promise(function(resolve, reject) {\r\n      // on success\r\n      resolve(value);\r\n  \r\n      // on failure\r\n      reject(reason);\r\n    });\r\n  \r\n    promise.then(function(value) {\r\n      // on fulfillment\r\n    }, function(reason) {\r\n      // on rejection\r\n    });\r\n    ```\r\n  \r\n    Advanced Usage:\r\n    ---------------\r\n  \r\n    Promises shine when abstracting away asynchronous interactions such as\r\n    `XMLHttpRequest`s.\r\n  \r\n    ```js\r\n    function getJSON(url) {\r\n      return new Promise(function(resolve, reject){\r\n        let xhr = new XMLHttpRequest();\r\n  \r\n        xhr.open('GET', url);\r\n        xhr.onreadystatechange = handler;\r\n        xhr.responseType = 'json';\r\n        xhr.setRequestHeader('Accept', 'application/json');\r\n        xhr.send();\r\n  \r\n        function handler() {\r\n          if (this.readyState === this.DONE) {\r\n            if (this.status === 200) {\r\n              resolve(this.response);\r\n            } else {\r\n              reject(new Error('getJSON: `' + url + '` failed with status: [' + this.status + ']'));\r\n            }\r\n          }\r\n        };\r\n      });\r\n    }\r\n  \r\n    getJSON('/posts.json').then(function(json) {\r\n      // on fulfillment\r\n    }, function(reason) {\r\n      // on rejection\r\n    });\r\n    ```\r\n  \r\n    Unlike callbacks, promises are great composable primitives.\r\n  \r\n    ```js\r\n    Promise.all([\r\n      getJSON('/posts'),\r\n      getJSON('/comments')\r\n    ]).then(function(values){\r\n      values[0] // => postsJSON\r\n      values[1] // => commentsJSON\r\n  \r\n      return values;\r\n    });\r\n    ```\r\n  \r\n    @class Promise\r\n    @param {Function} resolver\r\n    Useful for tooling.\r\n    @constructor\r\n  */\n\n\n  var Promise$1 = function () {\n    function Promise(resolver) {\n      this[PROMISE_ID] = nextId();\n      this._result = this._state = undefined;\n      this._subscribers = [];\n\n      if (noop !== resolver) {\n        typeof resolver !== 'function' && needsResolver();\n        this instanceof Promise ? initializePromise(this, resolver) : needsNew();\n      }\n    }\n    /**\r\n    The primary way of interacting with a promise is through its `then` method,\r\n    which registers callbacks to receive either a promise's eventual value or the\r\n    reason why the promise cannot be fulfilled.\r\n     ```js\r\n    findUser().then(function(user){\r\n      // user is available\r\n    }, function(reason){\r\n      // user is unavailable, and you are given the reason why\r\n    });\r\n    ```\r\n     Chaining\r\n    --------\r\n     The return value of `then` is itself a promise.  This second, 'downstream'\r\n    promise is resolved with the return value of the first promise's fulfillment\r\n    or rejection handler, or rejected if the handler throws an exception.\r\n     ```js\r\n    findUser().then(function (user) {\r\n      return user.name;\r\n    }, function (reason) {\r\n      return 'default name';\r\n    }).then(function (userName) {\r\n      // If `findUser` fulfilled, `userName` will be the user's name, otherwise it\r\n      // will be `'default name'`\r\n    });\r\n     findUser().then(function (user) {\r\n      throw new Error('Found user, but still unhappy');\r\n    }, function (reason) {\r\n      throw new Error('`findUser` rejected and we're unhappy');\r\n    }).then(function (value) {\r\n      // never reached\r\n    }, function (reason) {\r\n      // if `findUser` fulfilled, `reason` will be 'Found user, but still unhappy'.\r\n      // If `findUser` rejected, `reason` will be '`findUser` rejected and we're unhappy'.\r\n    });\r\n    ```\r\n    If the downstream promise does not specify a rejection handler, rejection reasons will be propagated further downstream.\r\n     ```js\r\n    findUser().then(function (user) {\r\n      throw new PedagogicalException('Upstream error');\r\n    }).then(function (value) {\r\n      // never reached\r\n    }).then(function (value) {\r\n      // never reached\r\n    }, function (reason) {\r\n      // The `PedgagocialException` is propagated all the way down to here\r\n    });\r\n    ```\r\n     Assimilation\r\n    ------------\r\n     Sometimes the value you want to propagate to a downstream promise can only be\r\n    retrieved asynchronously. This can be achieved by returning a promise in the\r\n    fulfillment or rejection handler. The downstream promise will then be pending\r\n    until the returned promise is settled. This is called *assimilation*.\r\n     ```js\r\n    findUser().then(function (user) {\r\n      return findCommentsByAuthor(user);\r\n    }).then(function (comments) {\r\n      // The user's comments are now available\r\n    });\r\n    ```\r\n     If the assimliated promise rejects, then the downstream promise will also reject.\r\n     ```js\r\n    findUser().then(function (user) {\r\n      return findCommentsByAuthor(user);\r\n    }).then(function (comments) {\r\n      // If `findCommentsByAuthor` fulfills, we'll have the value here\r\n    }, function (reason) {\r\n      // If `findCommentsByAuthor` rejects, we'll have the reason here\r\n    });\r\n    ```\r\n     Simple Example\r\n    --------------\r\n     Synchronous Example\r\n     ```javascript\r\n    let result;\r\n     try {\r\n      result = findResult();\r\n      // success\r\n    } catch(reason) {\r\n      // failure\r\n    }\r\n    ```\r\n     Errback Example\r\n     ```js\r\n    findResult(function(result, err){\r\n      if (err) {\r\n        // failure\r\n      } else {\r\n        // success\r\n      }\r\n    });\r\n    ```\r\n     Promise Example;\r\n     ```javascript\r\n    findResult().then(function(result){\r\n      // success\r\n    }, function(reason){\r\n      // failure\r\n    });\r\n    ```\r\n     Advanced Example\r\n    --------------\r\n     Synchronous Example\r\n     ```javascript\r\n    let author, books;\r\n     try {\r\n      author = findAuthor();\r\n      books  = findBooksByAuthor(author);\r\n      // success\r\n    } catch(reason) {\r\n      // failure\r\n    }\r\n    ```\r\n     Errback Example\r\n     ```js\r\n     function foundBooks(books) {\r\n     }\r\n     function failure(reason) {\r\n     }\r\n     findAuthor(function(author, err){\r\n      if (err) {\r\n        failure(err);\r\n        // failure\r\n      } else {\r\n        try {\r\n          findBoooksByAuthor(author, function(books, err) {\r\n            if (err) {\r\n              failure(err);\r\n            } else {\r\n              try {\r\n                foundBooks(books);\r\n              } catch(reason) {\r\n                failure(reason);\r\n              }\r\n            }\r\n          });\r\n        } catch(error) {\r\n          failure(err);\r\n        }\r\n        // success\r\n      }\r\n    });\r\n    ```\r\n     Promise Example;\r\n     ```javascript\r\n    findAuthor().\r\n      then(findBooksByAuthor).\r\n      then(function(books){\r\n        // found books\r\n    }).catch(function(reason){\r\n      // something went wrong\r\n    });\r\n    ```\r\n     @method then\r\n    @param {Function} onFulfilled\r\n    @param {Function} onRejected\r\n    Useful for tooling.\r\n    @return {Promise}\r\n    */\n\n    /**\r\n    `catch` is simply sugar for `then(undefined, onRejection)` which makes it the same\r\n    as the catch block of a try/catch statement.\r\n    ```js\r\n    function findAuthor(){\r\n    throw new Error('couldn't find that author');\r\n    }\r\n    // synchronous\r\n    try {\r\n    findAuthor();\r\n    } catch(reason) {\r\n    // something went wrong\r\n    }\r\n    // async with promises\r\n    findAuthor().catch(function(reason){\r\n    // something went wrong\r\n    });\r\n    ```\r\n    @method catch\r\n    @param {Function} onRejection\r\n    Useful for tooling.\r\n    @return {Promise}\r\n    */\n\n\n    Promise.prototype.catch = function _catch(onRejection) {\n      return this.then(null, onRejection);\n    };\n    /**\r\n      `finally` will be invoked regardless of the promise's fate just as native\r\n      try/catch/finally behaves\r\n    \r\n      Synchronous example:\r\n    \r\n      ```js\r\n      findAuthor() {\r\n        if (Math.random() > 0.5) {\r\n          throw new Error();\r\n        }\r\n        return new Author();\r\n      }\r\n    \r\n      try {\r\n        return findAuthor(); // succeed or fail\r\n      } catch(error) {\r\n        return findOtherAuther();\r\n      } finally {\r\n        // always runs\r\n        // doesn't affect the return value\r\n      }\r\n      ```\r\n    \r\n      Asynchronous example:\r\n    \r\n      ```js\r\n      findAuthor().catch(function(reason){\r\n        return findOtherAuther();\r\n      }).finally(function(){\r\n        // author was either found, or not\r\n      });\r\n      ```\r\n    \r\n      @method finally\r\n      @param {Function} callback\r\n      @return {Promise}\r\n    */\n\n\n    Promise.prototype.finally = function _finally(callback) {\n      var promise = this;\n      var constructor = promise.constructor;\n\n      if (isFunction(callback)) {\n        return promise.then(function (value) {\n          return constructor.resolve(callback()).then(function () {\n            return value;\n          });\n        }, function (reason) {\n          return constructor.resolve(callback()).then(function () {\n            throw reason;\n          });\n        });\n      }\n\n      return promise.then(callback, callback);\n    };\n\n    return Promise;\n  }();\n\n  Promise$1.prototype.then = then;\n  Promise$1.all = all;\n  Promise$1.race = race;\n  Promise$1.resolve = resolve$1;\n  Promise$1.reject = reject$1;\n  Promise$1._setScheduler = setScheduler;\n  Promise$1._setAsap = setAsap;\n  Promise$1._asap = asap;\n  /*global self*/\n\n  function polyfill() {\n    var local = void 0;\n\n    if (typeof global !== 'undefined') {\n      local = global;\n    } else if (typeof self !== 'undefined') {\n      local = self;\n    } else {\n      try {\n        local = Function('return this')();\n      } catch (e) {\n        throw new Error('polyfill failed because global object is unavailable in this environment');\n      }\n    }\n\n    var P = local.Promise;\n\n    if (P) {\n      var promiseToString = null;\n\n      try {\n        promiseToString = Object.prototype.toString.call(P.resolve());\n      } catch (e) {// silently ignored\n      }\n\n      if (promiseToString === '[object Promise]' && !P.cast) {\n        return;\n      }\n    }\n\n    local.Promise = Promise$1;\n  } // Strange compat..\n\n\n  Promise$1.polyfill = polyfill;\n  Promise$1.Promise = Promise$1;\n  return Promise$1;\n});","map":null,"metadata":{},"sourceType":"script"}